### PR TITLE
Create a pager component and add into contactList

### PR DESCRIPTION
**Pull request by JamesSCBD**

### CODE REVIEW

File: components/List/Pager.vue [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      currentPage: this.page, // Note: we can't modified props directly
+      perPage    : this.pageSize // Note: we can't modified props directly
+    }
+  },
+  computed: {
+  },
+  watch: {
+    currentPage (value){ this.$emit('update:page', value) },
+    perPage (value){ this.$emit('update:pageSize', value) },
+    syncQueryString: saveQueryString
```
Comment (sbilodeau) [inline components/List/Pager.vue:56]: What you implemented is: when the value `syncQueryString`  change you page and  skip to qs.

this is wrong. 

when the value of `syncQueryString` is true you have to:
 * load `page` & `pageSize` from the qs 
 * keep `page` & `pageSize` in sync into the query string when the value change 

Quote from the the trello card
> ## Behaviours
> if `syncQueryString` is `true` the component should load & save `pageSize` & `page` values to/from the queryString.

Comment (sbilodeau) [reply components/List/Pager.vue:56]: You make need this watch to start the sync if the use update afterward. but it wil have a different behaviour.

if true => false: delete the QS parameters
if false => true: save to QS

File: components/List/Pager.vue [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  },
+  mounted,
+  methods: { loadQueryString }
+}
+
+function mounted (){
+  this.loadQueryString()
```
Comment (sbilodeau) [inline components/List/Pager.vue:63]: Only load if is syncQueryString==true.

File: components/List/Pager.vue [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+function saveQueryString (){
+  if (this.syncQueryString){
+    const params = {
+      pageSize: undefined,
+      page    : undefined,
+      sorts   : undefined
+    }
+
+    if (this.page != defaultPage)     params.page     = this.page
+    if (this.pageSize    != defaultPageSize) params.pageSize = this.pageSize
```
Comment (sbilodeau) [inline components/List/Pager.vue:83]: code is not well aligned

File: components/List/Pager.vue [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      pageOptions,
+      currentPage: this.page, // Note: we can't modified props directly
+      perPage    : this.pageSize // Note: we can't modified props directly
+    }
+  },
+  computed: {
+  },
+  watch: {
+    currentPage (value){ this.$emit('update:page', value) },
+    perPage (value){ this.$emit('update:pageSize', value) },
```
Comment (sbilodeau) [inline components/List/Pager.vue:55]: avoid watches and use computed setter

File: components/List/Pager.vue [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+function loadQueryString (){
+  const { page, pageSize } = this.$route.query
+
+  this.currentPage = parseInt(page) >= 1 ?   parseInt(page) : defaultPage
+  this.perPage   = _.some(pageOptions, [ 'value', parseInt(pageSize) ])
+    ? parseInt(pageSize) : defaultPageSize
+}
+
+function saveQueryString (){
+  if (this.syncQueryString){
```
Comment (sbilodeau) [inline components/List/Pager.vue:75]: as `loadQueryString` `saveQueryString` should only do what it supposed to do:

*save value to QS*. the `if (this.syncQueryString)` should not be there

File: components/List/Pager.vue [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+const defaultPage     = 1
+
+export default {
+  name  : 'ContactsList',
+  mixins: [ mixin ],
+  props : {
+    pageSize       : { type: Number, default: 25 },
+    page           : { type: Number, default: 1 },
+    syncQueryString: { type: Boolean, default: false },
+    recordCount    : { type: Number, default: 0 }
```
Comment (sbilodeau) [inline components/List/Pager.vue:42]: add `validator` for props 
 * `pageSize `
 * `page `
 * `recordCount`.

https://vuejs.org/v2/guide/components-props.html#Prop-Validation

File: components/List/ContactsList.vue [hunk 7]
```diff
@@ -208,13 +186,12 @@ async function searchContacts (ctx){
...
     
     this.updateColumns()
-    this.saveQueryString()
 
     const query = this.buildQuery()
 
     if (!query)
       return
-      
+    this.syncQueryString = true
```
Comment (sbilodeau) [inline components/List/ContactsList.vue:194]: what is this? 

you have to decide if the pager will sync its state to the query string when you use it in the template.  the value should not change set syncQueryString. the the pager will update the query string of page/pagesize param only if true. NOT by toggling syncQueryString true/false

File: components/List/ContactsList.vue [hunk 8]
```diff
@@ -231,45 +208,10 @@ async function searchContacts (ctx){
   }
   finally {
     this.loading = false
+    this.syncQueryString = false
```
Comment (sbilodeau) [inline components/List/ContactsList.vue:211]: invalid use of syncQueryString. delete this line

File: components/List/ContactsList.vue [hunk 9]
```diff
@@ -1,19 +1,13 @@
...
-        prev-text="Prev"
-        next-text="Next"
-        aria-controls="KContactsList"
-        hide-goto-end-buttons
-        limit="4"
+      <pager
+        :page.sync="page"
+        :page-size.sync="pageSize"
+        :record-count="totalRows"
+        :sync-query-string="syncQueryString"
```
Comment (sbilodeau) [inline components/List/ContactsList.vue:8]: must be set to `true` or `false`. this value will not change over time

eg:
```html
:sync-query-string="true"
```

File: components/List/ContactsList.vue [hunk 10]
```diff
@@ -123,21 +109,21 @@ const registrationStatuses = {
...
-      pageOptions,
+      loading        : false,
+      columns        : [ ...baseColumns ],
+      totalRows      : 0,
+      page           : 1,
+      pageSize       : 25,
       registrationStatuses,
-      contacts   : []
+      contacts       : [],
+      syncQueryString: false
```
Comment (sbilodeau) [inline components/List/ContactsList.vue:126]: `syncQueryString` is useless delete

File: components/List/Pager.vue (hunk 1 above)
Comment (JamesSCBD) [reply components/List/Pager.vue]: `syncQueryString  ` is use only for Load & Save

File: components/List/Pager.vue (hunk 2 above)
Comment (JamesSCBD) [reply components/List/Pager.vue]: if `syncQueryString` Added at the `mounted`

File: components/List/Pager.vue [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+  this.perPage   = _.some(pageOptions, [ 'value', parseInt(pageSize) ])
+    ? parseInt(pageSize) : defaultPageSize
+}
+
+function saveQueryString (){
+  if (this.syncQueryString){
+    const params = {
+      pageSize: undefined,
+      page    : undefined,
+      sorts   : undefined
```
[the remark this answers is not in the document]
Comment (JamesSCBD) [reply components/List/Pager.vue]: Deleted `sorts`

File: components/List/Pager.vue (hunk 3 above)
Comment (JamesSCBD) [reply components/List/Pager.vue]: Done

File: components/List/Pager.vue (hunk 4 above)
Comment (JamesSCBD) [reply components/List/Pager.vue]: Added `currentPage `and `perPage `  declare  `computed` as You suggested

File: components/List/Pager.vue (hunk 5 above)
Comment (JamesSCBD) [reply components/List/Pager.vue]: Deleted `if (this.syncQueryString)` condition

**Comment by JamesSCBD** [comment]
> Please prioritize use computed properties. use watches only if absolutely needed

> Optional
> 4. `updateQueryStringIfRequired` (save Qs if sync is true) the one called from the currentpage / parpage setter

I have update QS from currentPage & perPage setter like this 

```
currentPage:
    {
      get (){ return this.page },
      set (value){
        this.$nextTick(() => {
          if (this.syncQueryString)
            this.saveQueryString()
        })
        this.$emit('update:page', value)
      }
    }
```
i have another solution we can use page & pageSize as a watcher like this
```
watch: {
    page (){
      if (this.syncQueryString)
        this.saveQueryString()
    },
    pageSize (){
      if (this.syncQueryString)
        this.saveQueryString()
    }
  }
```
But you suggest 
>Please prioritize use computed properties. use watches only if absolutely needed

So let me know which implementation is better. I will change accordingly

File: components/List/ContactsList.vue (hunk 10 above)
Comment (JamesSCBD) [reply components/List/ContactsList.vue]: Deleted

File: components/List/ContactsList.vue (hunk 9 above)
Comment (JamesSCBD) [reply components/List/ContactsList.vue]: Done

File: components/List/ContactsList.vue (hunk 8 above)
Comment (JamesSCBD) [reply components/List/ContactsList.vue]: Deleted

File: components/List/ContactsList.vue (hunk 7 above)
Comment (JamesSCBD) [reply components/List/ContactsList.vue]: Line Deleted

File: components/List/Pager.vue (hunk 6 above)
Comment (JamesSCBD) [reply components/List/Pager.vue]: I have added `validation ` props Let me know if i am added in right manner

File: components/controls/Pager.vue [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+      required: true,
+      default : false
+    },
+    recordCount:
+    {
+      type    : Number,
+      required: true,
+      default : 0,
+      validator (value){
+        return value > -1
```
Comment (sbilodeau) [inline components/controls/Pager.vue:69]: Use `value >=0`. -1 look strange and  -0.5 Is valid.

File: components/List/ContactsList.vue [hunk 13]
```diff
@@ -91,6 +85,7 @@
 <script>
 import _ from 'lodash'
 import { mapGetters, mapActions } from 'vuex'
+import pager from '../controls/Pager'
```
Comment (sbilodeau) [inline components/List/ContactsList.vue:88]: Use absolute reference when components do not goea along. 

`~/components/controls/pager.js`

Comment (sbilodeau) [reply components/List/ContactsList.vue:88]: Not sure if it’s `~` or `@`

File: components/List/Pager.vue [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+function mounted (){
+  if (this.syncQueryString)
+    this.loadQueryString()
+}
+
+function loadQueryString (){
+  const { page, pageSize } = this.$route.query
+
+  this.currentPage = parseInt(page) >= 1 ? parseInt(page) : defaultPage
+  this.perPage     = _.some(pageOptions, [ 'value', parseInt(pageSize) ]) ? parseInt(pageSize) : defaultPageSize
```
[the remark this answers is not in the document]
Comment (JamesSCBD) [reply components/List/Pager.vue]: I will make sure

File: components/controls/Pager.vue (hunk 12 above)
Comment (JamesSCBD) [reply components/controls/Pager.vue]: Done

File: components/List/ContactsList.vue (hunk 13 above)
Comment (JamesSCBD) [reply components/List/ContactsList.vue]: Done! Both are valid

File: components/controls/Pager.vue [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+      required: true,
+      default : false
+    },
+    recordCount:
+    {
+      type    : Number,
+      required: true,
+      default : 0,
+      validator (value){
+        return value > -0.5
```
Comment (sbilodeau) [inline components/controls/Pager.vue:69]: What is this: ` value > -0.5 `????

Value must be greater or equal to 0
the test should be 

```suggestion
        return value >= 0
```

File: components/List/ContactsList.vue [hunk 16]
```diff
@@ -1,19 +1,13 @@
...
-        prev-text="Prev"
-        next-text="Next"
-        aria-controls="KContactsList"
-        hide-goto-end-buttons
-        limit="4"
+      <pager
+        :page.sync="page"
+        :page-size.sync="pageSize"
+        :record-count="totalRows"
+        :sync-query-string="true"
```
Comment (sbilodeau) [inline components/List/ContactsList.vue:8]: The pageSize selector is taking the whole screen. sreenshot from trello [page-size-selector.png](https://trello-attachments.s3.amazonaws.com/56040c7949e3d23d65d8fc61/5e8e3624c6cc1a3eb48dd908/6b8029358f03c65f244051321718684c/page-size-selector.png)

File: components/controls/Pager.vue (hunk 15 above)
Comment (sbilodeau) [reply components/controls/Pager.vue:69]: sorry I read my previous comment. it was not wrriten correctly . you should hav read:
> Use value >=0. -1 look strange and **makes** -0.5 valid which is **wrong**.

**Pull request merged**